packages/remix-react/single-fetch.ts

```diff
@@ -0,0 +1,203 @@
+import type { DataStrategyMatch, ErrorResponse } from "@remix-run/router";
+import {
+  DecodedResponse,
+  UNSAFE_ErrorResponseImpl as ErrorResponseImpl,
+  redirect,
+} from "@remix-run/router";
+import type { DataStrategyFunctionArgs } from "react-router-dom";
+import { decode } from "turbo-stream";
+
+import { createRequestInit } from "./data";
+import type { AssetsManifest } from "./entry";
+import invariant from "./invariant";
+import type { RouteModules } from "./routeModules";
+
+type SingleFetchResult =
+  | { data: unknown; status?: number } // status only included in actions
+  | { error: unknown }
+  | { redirect: string; status: number; revalidate: boolean; reload: boolean };
+type SingleFetchResults = {
+  [key: string]: SingleFetchResult;
+};
+
+export function getSingleFetchDataStrategy(
+  manifest: AssetsManifest,
+  routeModules: RouteModules
+) {
+  return async ({ request, matches }: DataStrategyFunctionArgs) => {
+    // This function is the way for a loader/action to "talk" to the server
+    let singleFetch: (routeId: string) => Promise<unknown>;
+    if (request.method !== "GET") {
+      // Actions are simple since they're singular - just hit the server
+      singleFetch = async (routeId) => {
+        let url = singleFetchUrl(request.url);
+        let init = await createRequestInit(request);
+        let result = await fetchAndDecode(url, init);
+        return unwrapSingleFetchResult(result as SingleFetchResult, routeId);
+      };
```

Comment:
Action implementation.

Below is the more complex loader implementation which lets all loaders latch onto a single promise for the call.